doc/op_help.1.in
.TH OP_HELP 1 "@DATE@" "oprofile @VERSION@"
.UC 4
.SH NAME
op_help \- list OProfile events
.SH SYNOPSIS
.br
.B op_help
[
.I options
]
[event name]
.SH DESCRIPTION

By default,
.B op_help
lists the available performance counter options. If you
give it a symbolic event name, it will return the hardware
value (e.g. "op_help DATA_MEM_REFS").

.SH OPTIONS
.TP
.BI "--help / -? / --usage"
Show help message.
.br
.TP
.BI "--version / -v"
Show version.
.br
.TP
.BI "--cpu-type / -c"
Show the events for the given numerical CPU type.
.br
.TP
.BI "--get-cpu-type / -r"
Show the symbolic CPU name.
.br
.TP

.SH ENVIRONMENT
No special environment variables are recognised by op_help.

.SH FILES
.TP
.I $prefix/share/oprofile/
Event description files used by OProfile.
.TP
.I /var/lib/oprofile/samples/
The location of the generated sample files.

.SH VERSION
.TP
This man page is current for @PACKAGE@-@VERSION@.

.SH SEE ALSO
.BR oprofile(1)


ChangeLog
2003-07-05  Philippe Elie  <phil.el@wanadoo.fr>

	* daemon/opd_image.c:
	* libdb/db_insert.c: 64 bits arch compile fix, thanks to Falk Hueffner
	 <falk@debian.org>

2003-07-02  Philippe Elie  <phil.el@wanadoo.fr>

	* pp/opreport.cpp: honor options::show_header

2003-06-30  Graydon Hoare  <graydon@redhat.com>

	* libpp/op_header.cpp: Handle printing multiple bitmask values.

2003-06-19  John Levon  <levon@movementarian.org>

	* doc/Makefile.am: fix make distcheck

2003-06-18  John Levon  <levon@movementarian.org>

	* pp/opannotate_options.c: hide the unhandled
	--base-dir option

2003-06-18  John Levon  <levon@movementarian.org>

	* pp/Makefile.am: don't build opdiff

2003-06-17  John Levon  <levon@movementarian.org>

	* utils/opcontrol: don't accept "-r" - it's too easy to typo

2003-06-17  Will Cohen  <wcohen@redhat.com>

	* daemon/opd_cookie.h: Add s390 syscall value.
	
2003-06-17  John Levon  <levon@movementarian.org>

	* utils/opcontrol: don't accept "-r" - it's too easy to typo

2003-06-17  John Levon  <levon@movementarian.org>

	* doc/srcdoc/Doxyfile: removed

	* doc/srcdoc/Doxyfile.in:
	* configure.in: generate doxygen version nr. automatically

2003-06-17  Philippe Elie  <phil.el@wanadoo.fr>

	* libdb/db_insert.cpp: missing initialization of error message on
	  error path

2003-06-17  John Levon  <levon@movementarian.org>

	* utils/opcontrol: accept some short forms for common
	operations

2003-06-17  John Levon  <levon@movementarian.org>

	* utils/opcontrol: accept --foo blah as well as
	--foo=blah

2003-06-16  John Levon  <levon@movementarian.org>

	* utils/opcontrol: "opcontrol" will show usage
	instead of doing nothing

2003-06-16  Will Cohen  <wcohen@nc.rr.com>

	* libabi/abi.cpp:
	* libabi/op_import.cpp: Add required include for cassert.

2003-06-16  John Levon  <levon@movementarian.org>

	* events/x86-64.hammer.unit_masks: remove some
	bogus FIXMEs

2003-06-15  John Levon  <levon@movementarian.org>

	* README: fix autogen.sh line

2003-06-15  John Levon  <levon@movementarian.org>

	* daemon/opd_image.c: 64 bit fixes

2003-06-15  John Levon  <levon@movementarian.org>

	* utils/op_help.c: c89 fix

2003-06-15  John Levon  <levon@movementarian.org>

	* daemon/opd_image.c: c89 fix

2003-06-15  John Levon  <levon@movementarian.org>

	* doc/oprofile.xml: document --event

	* utils/opcontrol:
	* utils/op_help.c: allow unit mask, kernel, user to be
	optional. Verify not too many events are passed.

2003-06-15  John Levon  <levon@movementarian.org>

	* libop/op_cpu_type.c: fix nr_counters cpu type check

	* utils/op_help: implement --check_events, event validation
	and counter allocation

	* utils/opcontrol: use the above, implement --event=default

2003-06-14  John Levon  <levon@movementarian.org>

	* daemon/opd_image.c: fix a stupid error in my previous TGID patch

2003-06-12  Graydon Hoare  <graydon@redhat.com>

	* libop/op_events.c (op_check_events): Check bitmasks as well as 
	exclusive values, for unit masks.

2003-06-12  John Levon  <levon@movementarian.org>

	* daemon/opd_image.c: we were not handling a truncated
	TGID entry properly

2003-06-12  John Levon  <levon@movementarian.org>

	* libutil++/unique_storage.h: missing typename

2003-06-11  Philippe Elie  <phil.el@wanadoo.fr>

	* pp/opannotate_options.cpp: typo in option name

2003-06-11  Philippe Elie  <phil.el@wanadoo.fr>

	* libutil++/op_bfd.cpp: boring_symbol() new to get a better choice
	when eliminatiing symbol at identical vma

2003-06-11  Philippe Elie  <phil.el@wanadoo.fr>

	* utils/opcontrol: fix rtc option checking, bug added after 0.5.3,
	no big deal

2003-06-11  Philippe Elie  <phil.el@wanadoo.fr>

	* libregex/op_regex.cpp:
	* libpp/profile.cpp: compile fix, tree was broken by my previous patch

2003-06-11  Jason Yeh  <jason.yeh@amd.com>

	* events/i386.athlon.events: add CPU_CLK_UNHALTED

	* utils/opcontrol: use the above by default

	* events/x86-64.hammer.events:
	* events/x86-64.hammer.unit_masks: various fixes

2003-06-11  Philippe Elie  <phil.el@wanadoo.fr>

	* libpp/op_header.cpp:
	* libpp/profile.cpp:
	* libpp/profile_spec.cpp:
	* libpp/symbol_sort.cpp:
	* libregex/op_regex.cpp:
	* libutil++/op_bfd.cpp: use exception rather exit(EXIT_FAILURE) in
	library code

2003-06-11  John Levon  <levon@movementarian.org>

	* daemon/opd_image.c: future-proof the code handlers

2003-06-11  Philippe Elie  <phil.el@wanadoo.fr>

	* pp/opannotate.cpp: output_asm() avoid output when the set of selected
	symbols doesn't contain any samples

2003-06-11  Philippe Elie  <phil.el@wanadoo.fr>

	* libutil++/unique_storage.h: we don't need to check if key is present
	  before trying to insert it

2003-06-10  Philippe Elie  <phil.el@wanadoo.fr>

	* libpp/name_storage.h: minor tidy
	* libutil++/unique_storage.h: small efficiency improvement

2003-06-09  John Levon  <levon@movementarian.org>

	* pp/Makefile.am:
	* libutil/Makefile.am: fix make distcheck

2003-06-09  John Levon  <levon@movementarian.org>

	* libpp/opp_symbol.h: rename to ...

	* libpp/symbol.h: ... this

	* libpp/Makefile.am:
	* libpp/format_output.h:
	* libpp/format_output.cpp:
	* libpp/profile_container.h:
	* libpp/profile_container.cpp:
	* libpp/sample_container.h:
	* libpp/sample_container.cpp:
	* libpp/symbol_functors.h:
	* libpp/symbol_sort.cpp: fix up from above and some header
	cleanups

2003-06-09  John Levon  <levon@movementarian.org>

	* libpp/profile.h: std::pair not pair

2003-06-08  Philippe Elie  <phil.el@wanadoo.fr>

	* libpp/profile.cpp: don't underflow start offset when the vma range
	  is unknown
	* pp/opreport.cpp: use samples_range() not samples_range(0, ~0)

2003-06-08  Philippe Elie  <phil.el@wanadoo.fr>

	* libpp/profile.h:
	* libpp/profile.cpp: replace accumulate_samples() API by an iterator
	  interface
	* libpp/profile_container.h:
	* libpp/profilecontainer.cpp:
	* pp/opreport.cpp: use new API

2003-06-08  Philippe Elie  <phil.el@wanadoo.fr>

	* include/sstream: replace old version by the once shipped with

utils/op_help.c
/* $Id$ */
/* COPYRIGHT (C) 2000 THE VICTORIA UNIVERSITY OF MANCHESTER and John Levon
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the Free
 * Software Foundation; either version 2 of the License, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along with
 * this program; if not, write to the Free Software Foundation, Inc., 59 Temple
 * Place - Suite 330, Boston, MA 02111-1307, USA.
 */

/* Adapted from libpperf 0.5 by M. Patrick Goda and Michael S. Warren */

/* See IA32 Vol. 3 Appendix A */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <limits.h>

#include "version.h"
#include "op_events.h"
#include "op_events_desc.h"
#include "op_popt.h"

static op_cpu cpu_type = CPU_NO_GOOD;

/**
 * help_for_event - output event name and description
 * @i: event number
 *
 * output an help string for the event @i
 */
static void help_for_event(int i)
{
	uint j,k;
	op_cpu cpu;
	uint mask;

	printf("%s", op_events[i].name);

	printf(": (counter: ");
	if (op_events[i].counter_mask == CTR_ALL) {
		printf("all");
	} else {
		mask = op_events[i].counter_mask;
		for (k = 0; k < CHAR_BIT * sizeof(op_events[i].counter_mask); ++k) {
			if (mask & (1 << k)) {
				printf("%d", k);
				mask &= ~(1 << k);
				if (mask)
					printf(", ");
			}
		}
	}
	printf(")");

	printf(" (supported cpu: ");
	mask = op_events[i].cpu_mask;
	for (cpu = 0; cpu < MAX_CPU_TYPE; ++cpu) {
		if (mask & (1 << cpu)) {
			printf("%s", op_get_cpu_type_str(cpu));
			mask &= ~(1 << cpu);
			if (mask)
				printf(", ");
		}
	}

	printf(")\n\t%s (min count: %d)\n", op_event_descs[i], op_events[i].min_count);

	if (op_events[i].unit) {
		int unit_idx = op_events[i].unit;

		printf("\tUnit masks\n");
		printf("\t----------\n");

		for (j=0; j < op_unit_masks[unit_idx].num; j++) {
			printf("\t%.2x: %s\n",
			       op_unit_masks[unit_idx].um[j],
			       op_unit_descs[unit_idx].desc[j]);
		}
	}
}

static int showvers;
static int get_cpu_type;
static const char * event_name;

static struct poptOption options[] = {
	{ "cpu-type", 'c', POPT_ARG_INT, &cpu_type, 0,
	  "force cpu type", "cpu type", },
	{ "get-cpu-type", 'r', POPT_ARG_NONE, &get_cpu_type, 0,
	  "show the auto detected cpu type", NULL, },
	{ "version", 'v', POPT_ARG_NONE, &showvers, 0, "show version", NULL, },
	POPT_AUTOHELP
	{ NULL, 0, 0, NULL, 0, NULL, NULL, },
};

/**
 * get_options - process command line
 * @argc: program arg count
 * @argv: program arg array
 *
 * Process the arguments, fatally complaining on error.
 */
static void get_options(int argc, char const * argv[])
{
	poptContext optcon;

	optcon = op_poptGetContext(NULL, argc, argv, options, 0);

	if (showvers) {
		show_version(argv[0]);
	}

	/* non-option, must be a valid event name */
	event_name = poptGetArg(optcon);

	poptFreeContext(optcon);
}

int main(int argc, char const *argv[])
{
	int cpu_type_mask;
	uint j;

	cpu_type = op_get_cpu_type();

	get_options(argc, argv);

	if (cpu_type < 0 || cpu_type >= MAX_CPU_TYPE) {
		fprintf(stderr, "invalid cpu type %d !\n", cpu_type);
		exit(EXIT_FAILURE);
	}

	if (get_cpu_type) {
		printf("%d\n", cpu_type);
		exit(EXIT_SUCCESS);
	}

	if (event_name) {
		cpu_type_mask = 1 << cpu_type;
		for (j=0; j < op_nr_events; j++) {
			if (!strcmp(op_events[j].name, event_name) && 
			    (op_events[j].cpu_mask & cpu_type_mask)) {
				printf("%d\n", op_events[j].val); 
				exit(EXIT_SUCCESS);
			}
		}
		fprintf(stderr, "No such event \"%s\"\n", event_name);
		exit(EXIT_FAILURE);
	} 

	printf("oprofile: available events\n");
	printf("--------------------------\n\n");
	if (cpu_type == CPU_ATHLON)
		printf ("See AMD document x86 optimisation guide (22007.pdf), Appendix D\n\n");
	else
		printf("See Intel Architecture Developer's Manual\nVol. 3 (), Appendix A\n\n");

	cpu_type_mask = 1 << cpu_type;
	for (j = 0; j < op_nr_events; j++) {
		if ((op_events[j].cpu_mask & cpu_type_mask) != 0) {
			help_for_event(j);
		}
	}

	return EXIT_SUCCESS;
}
